From 32c4e875cda858b68acb5dc2b03882af0e29f530 Mon Sep 17 00:00:00 2001
From: CIQ Kernel Automation <ciq_kernel_automation@ciq.com>
Date: Mon, 15 Jun 2026 08:16:44 +0000
Subject: [PATCH] kprobes/x86: Use copy_from_kernel_nofault() to read from
 unsafe address

jira VULN-37443
cve CVE-2024-26946
commit-author Masami Hiramatsu (Google) <mhiramat@kernel.org>
commit 4e51653d5d871f40f1bd5cf95cc7f2d8b33d063b

Read from an unsafe address with copy_from_kernel_nofault() in
arch_adjust_kprobe_addr() because this function is used before checking
the address is in text or not. Syzcaller bot found a bug and reported
the case if user specifies inaccessible data area,
arch_adjust_kprobe_addr() will cause a kernel panic.

[ mingo: Clarified the comment. ]

Fixes: cc66bb914578 ("x86/ibt,kprobes: Cure sym+0 equals fentry woes")
	Reported-by: Qiang Zhang <zzqq0103.hey@gmail.com>
	Tested-by: Jinghao Jia <jinghao7@illinois.edu>
	Signed-off-by: Masami Hiramatsu (Google) <mhiramat@kernel.org>
	Signed-off-by: Ingo Molnar <mingo@kernel.org>
Link: https://lore.kernel.org/r/171042945004.154897.2221804961882915806.stgit@devnote2
(cherry picked from commit 4e51653d5d871f40f1bd5cf95cc7f2d8b33d063b)
	Signed-off-by: CIQ Kernel Automation <ciq_kernel_automation@ciq.com>
---
 arch/x86/kernel/kprobes/core.c | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/arch/x86/kernel/kprobes/core.c b/arch/x86/kernel/kprobes/core.c
index 7c4ab8870da4..da22383b1eff 100644
--- a/arch/x86/kernel/kprobes/core.c
+++ b/arch/x86/kernel/kprobes/core.c
@@ -299,7 +299,16 @@ static int can_probe(unsigned long paddr)
 kprobe_opcode_t *arch_adjust_kprobe_addr(unsigned long addr, unsigned long offset,
 					 bool *on_func_entry)
 {
-	if (is_endbr(*(u32 *)addr)) {
+	u32 insn;
+
+	/*
+	 * Since 'addr' is not guaranteed to be safe to access, use
+	 * copy_from_kernel_nofault() to read the instruction:
+	 */
+	if (copy_from_kernel_nofault(&insn, (void *)addr, sizeof(u32)))
+		return NULL;
+
+	if (is_endbr(insn)) {
 		*on_func_entry = !offset || offset == 4;
 		if (*on_func_entry)
 			offset = 4;
-- 
2.52.0

